### ISSUE TITLE
Deprecate digital_ocean module

### ISSUE THREAD

**Issue by kontrafiktion** [ansible/ansible-modules-core#4219]
##### ISSUE TYPE
- Feature Idea
##### COMPONENT NAME

digital_ocean module
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file =
  configured module search path = Default w/o overrides
```
##### SUMMARY

There is an overlap between "digital_ocean" and "digital_ocean_sshkey". The first allows to work with droplets _and_ ssh keys, the latter only handles ssh keys.
I propose to deprecate the existing digital_ocean module and create a new digital_ocean_droplet module. 
##### DETAILS
1. The new digital_ocean_droplet module should not use dopy but simply ansible.module_utils.urls
2. It must be ensured that all ssh features of the digital_ocean module are covered by the digital_ocean_sshkey module

**Comment by harneksidhu** [comment]
I agree that dopy should be removed as it is not very well maintained. However, instead of using ansible.module_utils.urls, do you think we could use https://github.com/koalalorenzo/python-digitalocean instead since it is maintained very well?

**Comment by bcoca** [comment]
I think either idea works for us, now to bikeshed: do_droplet as a name?

**Comment by harneksidhu** [comment]
I agree that dopy should be removed as it is not very well maintained. However, instead of using ansible.module_utils.urls, do you think we could use https://github.com/koalalorenzo/python-digitalocean instead since it is maintained very well?

**Comment by kontrafiktion** [comment]
@harneksidhu I would prefer to remove _any_ extra dependencies
@bcoca I like do_droplet, but then we would have to rename all modules. Kills backward compatibility – or is it possible to have an alias for a module?

**Comment by harneksidhu** [comment]
@kontrafiktion I would love to help with the development of this module if you need any assistance.

My main concern about only using ansible.module_utils.urls is that digital ocean appears to have a pipeline of features coming in (https://beta.digitalocean.com/). It would be hard to quickly add these new features onto Ansible on our own since there is already a large community of developers working on that initiative within python-digitalocean (they already started to work on block storage support as an example https://github.com/koalalorenzo/python-digitalocean/blob/master/digitalocean/Volume.py).

I am open to helping with any direction chosen of course!

**Comment by bcoca** [comment]
@kontrafiktion to alias a module just create a symlink with leading underscore `_alias`  that points to the new name (this was first feature i added as ansible employee!).

As for using the python library or not, this is a choice i'm leaving up to the community, I have barely looked at the DO API and don't know how complex/or not it is and how easy would it be to use directly or what lag the 'good library' has vs the features. 

We normally prefer using an established library, but in this case that seems to be in flux right now. We have been 'bitten' by either the API or library being in flux (docker , azure ...) so I know it is not going to be a smooth ride either way.

**Comment by jctanner** [comment]
@mgregson, @alukovenko ping

**Comment by alukovenko** [comment]
DO API is not really complex, and I would also like to remove external dependencies. @harneksidhu concern about upcoming features is valid; however, those features will still need to be included in Ansible module regardless of internal implementation - at the end of the day, we still need to update the module itself.

One additional concern about using external modules is that it uses separate versioning / branching, which can (and will) lead to bugs that are hard to reproduce.

Given the fact that DO API is quite straightforward, I would prefer to:
1. Create our own implementation of API
2. Create new modules according to @kontrafiktion  proposal
3. Deprecate existing modules, at the same time maintaining compatibility using _alias

We could use _do_droplet_ and _do_storage_ names for the new implementation

**Comment by harneksidhu** [comment]
Should we consolidate this with #4200 and include block storage support as well?

**Comment by alukovenko** [comment]
Let's merge #4200 to this issue

**Comment by alukovenko** [comment]
@kontrafiktion , @harneksidhu - it would be great do collaborate on this to add up the effort; would you be willing to contribute to developing this new module(s)?

**Comment by kontrafiktion** [comment]
I already started, I think I have a first version next week

**Comment by alukovenko** [comment]
@kontrafiktion would you like to do it by your own? I would like to join.

**Comment by kontrafiktion** [comment]
You are more than welcome. I have no idea yet, how to split up the work.
I just published my work in progress at https://github.com/kontrafiktion/ansible-modules-core/tree/4219_deprecate_digital_ocean_module

**Comment by harneksidhu** [comment]
@kontrafiktion Thank's for taking a shot at this! Not sure how to split up the work either but am available if you want to send some tasks my way or need help with testing/reviewing code. I'm looking forward to getting rid of dopy.

**Comment by alukovenko** [comment]
@kontrafiktion - would you be able to give us some access for your fork? then we can collaborate in a usual way with creating issues and committing directly...

**Comment by kontrafiktion** [comment]
@alukovenko not sure, what you mean: I can add you as a collaborator. Anybody else as well?

**Comment by alukovenko** [comment]
That's exactly what I meant. Thanks!

**Comment by alukovenko** [comment]
@kontrafiktion: Just a followup: if you turn on Options -> Features -> Issues, then we can use it to split tasks

**Comment by kontrafiktion** [comment]
done

**Comment by Fale** [comment]
DO has [deprecated v1](https://developers.digitalocean.com/documentation/v1/) and therefore digital_ocean_sshkey is now broken. I hope the rewrite will soon be completed and merged in Ansible before the next release

**Comment by kontrafiktion** [comment]
The digital_ocean module handles ssh keys as well

**Comment by ansibot** [comment]
@alukovenko, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@alukovenko, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by harneksidhu** [comment]
I should have a module submitted by Monday

**Comment by pmarques** [comment]
Hi guys, I'm think I'm a bit late, I wrote a library inside my playbooks because I needed to manage the Floating IPs. I started it some time ago and I only use the internal ansible modules, ansible fetch_url.
I made it a separated "role" just to open to the community, [https://github.com/pmarques/ansible-digitalocean](https://github.com/pmarques/ansible-digitalocean). So, let me know our thoughts about it.

**Comment by harneksidhu** [comment]
@pmarques Can you rename the module using a do_\* prefix and submit a PR? An example would be do_floating_ip or something similar. We will be renaming the modules using the prefixes. We can review the code together once you submitted it the PR.

**Comment by pmarques** [comment]
@harneksidhu let me know what do you think. ansible/ansible-modules-core#5221

There is one thing that I don't like, there is no clean way to test if a Floating IP is created [DigitalOcean community questions](https://www.digitalocean.com/community/questions/how-can-i-verify-if-a-floating-ip-is-created-i-need-an-id-or-something-deterministic-to-identify-it).

**Comment by andrewsomething** [comment]
Hi. I'm looking into porting the existing modules to not depend on dopy. (WIP here: https://github.com/andrewsomething/ansible-modules-core/tree/issues/4219)

The `digital_ocean_tag` module was already written using just `ansible.module_utils.urls.fetch_url` It contains two generic classes (`Rest` and `Response`) that would be helpful in this. What's the best practice to make them reusable? Create `ansible.module_utils.digital_ocean`? @pmarques is using a different approach in his Floating IPs branch. It seems like we should standardize on something across the various DigitalOcean modules.

**Comment by kontrafiktion** [comment]
We discussed exactly that question in this thread: https://github.com/kontrafiktion/ansible-modules-core/issues/1
For now we decided to first implement everything without a common module (simply using copy/paste). Only when we really know what should belong in the common module, we want to move forward. And the question even is: does this belong into a 'digital ocean' utils module or should it rather be part of the 'urls' module? 
And: ansible-modules-core is planned to be merged with ansible, so we wanted to wait fir that as well

**Comment by pmarques** [comment]
@andrewsomething I had already implemented a while ago, I didn't had seen the digital_ocean_tag module and so I made a simplest version :/ I was also thinking about create a common module in module_utils but like @kontrafiktion I would like to have it working/implemented before moving into there.

**Comment by harneksidhu** [comment]
@andrewsomething

Seems like we are already working on same thing.
Here's my work in progress.

https://github.com/harneksidhu/ansible-modules-core/blob/4219_do_rework/cloud/digital_ocean/do_droplet.py

Are you interested in doing the droplet module yourself or do you want me to complete it? I was planning on finishing it by Thursday. Let me know if i should continue

**Comment by andrewsomething** [comment]
@harneksidhu Awesome stuff. Please continue, but let me know if you aren't able to complete it for some reason and I can pick up where you left off.

**Comment by harneksidhu** [comment]
@andrewsomething

Thank you! I'll ping you for a PR review.

**Comment by pmarques** [comment]
I've the updated to use the same approach of Request / Response ansible/ansible-modules-core#5221

Note that it is a work in progress and maybe needs a cleanup

**Comment by pmarques** [comment]
I also have ported ssh keys to rid off _dopy_.... [do_sshkeys.py](https://github.com/pmarques/ansible-digitalocean/blob/master/library/do_sshkeys.py).

**Comment by harneksidhu** [comment]
@pmarques thanks for porting the ssh key module. I'll use it to test do_droplet that I'm still working on

**Comment by harneksidhu** [comment]
@pmarques 

Should this edge case be supported by do_sshkeys?
```
  - name: Store public ssh key in digital ocean
    do_sshkeys:
      state: present
      name: key1
      ssh_pub_key: "{{ public_key.stdout }}"
      oauth_token: "{{ digital_ocean_api_token }}"
    register: key1
  - debug: var=key1
  - name: Store public ssh key in digital ocean
    do_sshkeys:
      state: present
      name: key2
      ssh_pub_key: "{{ public_key.stdout }}"
      oauth_token: "{{ digital_ocean_api_token }}"
    register: key2
  - debug: var=key2
```

I expect the ssh key name to be renamed to key2 but it stays as key1.

**Comment by harneksidhu** [comment]
@pmarques Can do_sshkeys use the following keys for the api_token:

`api_token` with an alias of `API_TOKEN`

Can it also accept the following environment variables: 
`DO_API_TOKEN` or `DO_API_KEY`

It would be nice to have this so that it can be consistent with the other modules

**Comment by harneksidhu** [comment]
@pmarques 
Can I complete do_sshkeys.py on my branch? I am ready to submit a PR for do_droplet.py but I need do_sshkeys.py completed first in order for it to work since ssh key id is a requirement for the droplet module.

**Comment by pmarques** [comment]
@harneksidhu I'm updating it submit a PR, implementing the alias and add the rename. You can use the Key fingerprint as key ID

**Comment by pmarques** [comment]
@harneksidhu PR submitted, ansible/ansible-modules-core#5275 @andrewsomething if you could give some feedback on this one I appreciate :)

**Comment by harneksidhu** [comment]
Thanks!! :

**Comment by harneksidhu** [comment]
#5278 Has been submitted. Will need #5275 to go in first though.

**Comment by ansibot** [comment]
This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by ed11s8c** [comment]
Any updates on the progress of removing the requirement for DOPY with digital ocean modules?

**Comment by ansibot** [comment]
This issue was migrated to https://github.com/ansible/ansible/issues/29740
